deepspeed/runtime/engine.py

```diff
@@ -1754,9 +1759,13 @@ def allreduce_bucket(self, bucket, dp_group):
                     tensor_to_allreduce.mul_(self.gradient_predivide_factor() /
                                              dist.get_world_size(group=dp_group))
         else:
-            tensor_to_allreduce.div_(dist.get_world_size(group=dp_group))
+            tensor_to_allreduce.mul_(1. / dist.get_world_size(group=dp_group))
```

Comment:
Can you please explain the benefit of this change?

Reply:
Multiplication is 2-3 times faster than division. There we divide 1 number and after that multiply array to this value. This is faster than dividing the whole array to value.

```python
import time
import numpy as np

arr = np.random.randn(1024, 1024, 3).astype(np.float32)
val = 15648931
val_inverse = 1.0 / val
N = 1000
print(arr.shape, arr.min(), arr.max())

s = time.time()
for _ in range(N):
    res = arr / val
dt_div = time.time() - s

s = time.time()
for _ in range(N):
    res = arr * val_inverse
dt_mul = time.time() - s

print(f"dt_div: {dt_div:.3f} dt_mul: {dt_mul:.3f}")
# dt_div: 3.730 dt_mul: 1.734
```

Reply:
Thanks for the explanation. My recollection is that modern compilers automatically perform such optimizations :).

Reply:
> My recollection is that modern compilers automatically perform such optimizations

No, at first np.ndarray is complex data type and compiler do not know how to optimize overloaded operation.
And another point that in some cases division and multiplication to inverse number is not fully equivalent because float data type has some rounding errors.

Reply:
Fair enough. Thanks.